DM-33414: Allow id to be used in file template

timj opened the pull request:
@gpdf / @TallJimbo this does the job -- I also restrict the id to having to be in the filename itself if it is specified anywhere.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Allow id to be used in file template

If id is used it must at least be present in the file
name part of the template path.

TallJimbo commented on python/lsst/daf/butler/core/fileTemplates.py:
```diff
@@ -552,9 +558,11 @@ def format(self, ref: DatasetRef) -> str:
+        if not usedRun and not usedId:
+            missing = ("run" if not usedRun else None, "id" if not usedId else None)
+            text = " or ".join(f"'{m}'" for m in missing if m is not None)
+            raise KeyError(f"Template does not include {text}.")
```
Is there a similar check for the dataset type name that could also be relaxed?

timj replied:
It doesn't look like we require the dataset type to be in the template. "run" was the only thing we required from what I could see (and I was able to use a file template of just `{id}`.

timj replied:
The fact that this line has no coverage implies that we catch the no-run logic higher up and none of this is needed any more.

TallJimbo replied:
Sounds like we should just remove it, then, and update the docs to say it's expected?

timj replied:
I think that this block of code was an early attempt at template validation but at some point we moved the validation logic earlier and never noticed that this block was irrelevant. I'll remove it.

TallJimbo reviewed (approved):
Seems fine - functionality is clearly what we want, but I'm a little worried that the template validation code is getting a little hard to follow now that we have so many rules (especially if we don't really understand why some of the new lines aren't in the coverage report).

timj committed:
Remove unused old validation code in file templates

During DM-17025 the code in the format() method to check
that run was present was made irrelevant because of an
earlier check that was added as part of generic
template validation.

timj committed:
Add news fragment
